File: planet-examples/src/main/java/org/saar/minecraft/World.java
package org.saar.minecraft;

import org.saar.maths.transform.Position;
import org.saar.minecraft.generator.WorldGenerator;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.stream.Stream;

public class World {

    private final ExecutorService executorService;

    private final List<Chunk> chunks = new ArrayList<>();
    private final WorldGenerator generator;

    public World(WorldGenerator generator, int threads) {
        this.generator = generator;
        this.executorService = Executors.newFixedThreadPool(threads);
    }

    public static int worldToChunkCoordinate(int w) {
        return w >= 0 ? w / 16 : (w + 1) / 16 - 1;
    }

    public void addChunk(Chunk chunk) {
        getChunks().add(chunk);
    }

    public boolean hasChunk(int x, int z) {
        return getChunk(x, z) != EmptyChunk.getInstance();
    }

    public IChunk getChunk(int x, int z) {
        for (Chunk chunk : getChunks()) {
            if (chunk.getPosition().equals(x, z)) {
                return chunk;
            }
        }
        return EmptyChunk.getInstance();
    }

    public IChunk getChunk(Position position) {
        final int x = worldToChunkCoordinate((int) position.getX());
        final int z = worldToChunkCoordinate((int) position.getZ());
        return getChunk(x, z);
    }

    public BlockContainer getBlockContainer(Position position) {
        return getBlockContainer(position.getX(), position.getY(), position.getZ());
    }

    public BlockContainer getBlockContainer(float wx, float wy, float wz) {
        final int x = (int) Math.floor(wx);
        final int y = (int) Math.floor(wy);
        final int z = (int) Math.floor(wz);
        return new BlockContainer(x, y, z, getBlock(x, y, z));
    }

    public void setLight(int x, int y, int z, byte value) {
        final int cx = worldToChunkCoordinate(x);
        final int cz = worldToChunkCoordinate(z);
        final IChunk chunk = getChunk(cx, cz);
        if (chunk instanceof Chunk) {
            final int lx = x - cx * 16;
            final int lz = z - cz * 16;
            ((Chunk) chunk).setLight(lx, y, lz, value);
        }
    }

    public int getLight(int x, int y, int z) {
        final int cx = worldToChunkCoordinate(x);
        final int cz = worldToChunkCoordinate(z);
        final IChunk chunk = getChunk(cx, cz);

        final int lx = x - cx * 16;
        final int lz = z - cz * 16;
        return chunk.getLight(lx, y, lz);
    }

    public Block getBlock(Position position) {
        final int x = (int) Math.floor(position.getX());
        final int y = (int) Math.floor(position.getY());
        final int z = (int) Math.floor(position.getZ());
        return getBlock(x, y, z);
    }

    public Block getBlock(int x, int y, int z) {
        final int cx = worldToChunkCoordinate(x);
        final int cz = worldToChunkCoordinate(z);
        final IChunk chunk = getChunk(cx, cz);

        final int lx = x - cx * 16;
        final int lz = z - cz * 16;
        return chunk.getBlock(lx, y, lz);
    }

    public int getHeight(int x, int z) {
        final int cx = worldToChunkCoordinate(x);
        final int cz = worldToChunkCoordinate(z);
        final IChunk chunk = getChunk(cx, cz);

        final int lx = x - cx * 16;
        final int lz = z - cz * 16;
        return chunk.getHeight(lx, lz);
    }

    public void setBlock(int x, int y, int z, Block block) {
        final int cx = worldToChunkCoordinate(x);
        final int cz = worldToChunkCoordinate(z);
        final IChunk chunk = getChunk(cx, cz);

        final int lx = x - cx * 16;
        final int lz = z - cz * 16;
        chunk.setBlock(lx, y, lz, block);

        this.executorService.submit(() -> {
            chunk.updateMesh();
            getChunk(cx - 1, cz).updateMesh();
            getChunk(cx + 1, cz).updateMesh();
            getChunk(cx, cz - 1).updateMesh();
            getChunk(cx, cz + 1).updateMesh();
        });
    }

    public List<Chunk> getChunks() {
        return this.chunks;
    }

    public void delete() {
        for (Chunk chunk : getChunks()) {
            chunk.delete();
        }
        this.executorService.shutdown();
    }

    public CompletableFuture<Void> generateAround(Position position, int radius) {
        final int px = (int) position.getX();
        final int pz = (int) position.getZ();
        final int cx = worldToChunkCoordinate(px);
        final int cz = worldToChunkCoordinate(pz);

        final List<Chunk> toDelete = new ArrayList<>();
        for (Chunk chunk : getChunks()) {
            if (Math.abs(chunk.getPosition().x() - cx) > radius + 1
                    || Math.abs(chunk.getPosition().y() - cz) > radius + 1) {
                toDelete.add(chunk);
            }
        }
        toDelete.forEach(Chunk::delete);
        getChunks().removeAll(toDelete);

        final List<Chunk> chunks = new ArrayList<>();

        for (int x = cx - radius; x < cx + radius; x++) {
            for (int z = cz - radius; z < cz + radius; z++) {
                if (!hasChunk(x, z)) chunks.add(new Chunk(this, x, z));
            }
        }

        chunks.sort(Comparator.comparingInt(c -> (int) c.getPosition().distanceSquared(cx, cz)));

        getChunks().addAll(chunks);

        final Stream<CompletableFuture<Void>> futures = chunks.stream().map(chunk -> CompletableFuture
                .runAsync(() -> this.generator.generateChunk(chunk), this.executorService)
                .exceptionally(f -> {
                    System.err.println("Failed to generate chunk " +
                            chunk.getPosition().x() + ", " +
                            chunk.getPosition().y() + " ");
                    f.printStackTrace();
                    return null;
                }));

        return CompletableFuture.allOf(futures.toArray(CompletableFuture[]::new)).thenRun(() ->
                chunks.stream().flatMap(c -> Stream.of(c,
                        getChunk(c.getPosition().x(), c.getPosition().y() + 1),
                        getChunk(c.getPosition().x(), c.getPosition().y() - 1),
                        getChunk(c.getPosition().x() + 1, c.getPosition().y()),
                        getChunk(c.getPosition().x() - 1, c.getPosition().y()))
                ).distinct().forEach(IChunk::updateMesh));
    }
}

File: planet-examples/src/main/java/org/saar/minecraft/World.kt
package org.saar.minecraft

import org.saar.maths.transform.Position
import org.saar.maths.transform.ReadonlyPosition
import org.saar.minecraft.chunk.ChunkConstants
import org.saar.minecraft.generator.WorldGenerator
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.function.Function
import kotlin.math.abs
import kotlin.math.floor

class World(private val generator: WorldGenerator, threads: Int) {
    private val executorService: ExecutorService

    val chunks: MutableList<Chunk> = ArrayList<Chunk>()

    init {
        this.executorService = Executors.newFixedThreadPool(threads)
    }

    fun addChunk(chunk: Chunk) {
        this.chunks.add(chunk)
    }

    fun hasChunk(x: Int, z: Int): Boolean {
        return getChunk(x, z) !== EmptyChunk
    }

    fun getChunk(x: Int, z: Int): IChunk {
        for (chunk in this.chunks) {
            if (chunk.getPosition().equals(x, z)) {
                return chunk
            }
        }
        return EmptyChunk
    }

    fun getChunk(position: Position): IChunk {
        val x: Int = worldToChunkCoordinate(position.x.toInt())
        val z: Int = worldToChunkCoordinate(position.z.toInt())
        return getChunk(x, z)
    }

    fun getBlockContainer(position: Position): BlockContainer {
        return getBlockContainer(position.x, position.y, position.z)
    }

    fun getBlockContainer(wx: Float, wy: Float, wz: Float): BlockContainer {
        val x = floor(wx.toDouble()).toInt()
        val y = floor(wy.toDouble()).toInt()
        val z = floor(wz.toDouble()).toInt()
        return BlockContainer(x, y, z, getBlock(x, y, z))
    }

    fun setLight(x: Int, y: Int, z: Int, value: Byte) {
        val cx: Int = worldToChunkCoordinate(x)
        val cz: Int = worldToChunkCoordinate(z)
        val chunk = getChunk(cx, cz)
        if (chunk is Chunk) {
            val lx = x - cx * ChunkConstants.SIZE
            val lz = z - cz * ChunkConstants.SIZE
            chunk.setLight(lx, y, lz, value)
        }
    }

    fun getLight(x: Int, y: Int, z: Int): Int {
        val cx: Int = worldToChunkCoordinate(x)
        val cz: Int = worldToChunkCoordinate(z)
        val chunk = getChunk(cx, cz)

        val lx = x - cx * ChunkConstants.SIZE
        val lz = z - cz * ChunkConstants.SIZE
        return chunk.getLight(lx, y, lz)
    }

    fun getBlock(position: ReadonlyPosition): Block {
        val x = floor(position.getValue().x().toDouble()).toInt()
        val y = floor(position.getValue().y().toDouble()).toInt()
        val z = floor(position.getValue().z().toDouble()).toInt()
        return getBlock(x, y, z)
    }

    fun getBlock(x: Int, y: Int, z: Int): Block {
        val cx: Int = worldToChunkCoordinate(x)
        val cz: Int = worldToChunkCoordinate(z)
        val chunk = getChunk(cx, cz)

        val lx = x - cx * ChunkConstants.SIZE
        val lz = z - cz * ChunkConstants.SIZE
        return chunk.getBlock(lx, y, lz)
    }

    fun getHeight(x: Int, z: Int): Int {
        val cx: Int = worldToChunkCoordinate(x)
        val cz: Int = worldToChunkCoordinate(z)
        val chunk = getChunk(cx, cz)

        val lx = x - cx * ChunkConstants.SIZE
        val lz = z - cz * ChunkConstants.SIZE
        return chunk.getHeight(lx, lz)
    }

    fun setBlock(x: Int, y: Int, z: Int, block: Block) {
        val cx: Int = worldToChunkCoordinate(x)
        val cz: Int = worldToChunkCoordinate(z)
        val chunk = getChunk(cx, cz)

        val lx = x - cx * ChunkConstants.SIZE
        val lz = z - cz * ChunkConstants.SIZE
        chunk.setBlock(lx, y, lz, block)

        this.executorService.submit(Runnable {
            chunk.updateMesh()
            getChunk(cx - 1, cz).updateMesh()
            getChunk(cx + 1, cz).updateMesh()
            getChunk(cx, cz - 1).updateMesh()
            getChunk(cx, cz + 1).updateMesh()
        })
    }

    fun delete() {
        for (chunk in this.chunks) {
            chunk.delete()
        }
        this.executorService.shutdown()
    }

    fun generateAround(position: Position, radius: Int): CompletableFuture<Void> {
        val px = position.x.toInt()
        val pz = position.z.toInt()
        val cx: Int = worldToChunkCoordinate(px)
        val cz: Int = worldToChunkCoordinate(pz)

        val toDelete: MutableList<Chunk> = ArrayList<Chunk>()
        for (chunk in this.chunks) {
            if (abs(chunk.getPosition().x() - cx) > radius + 1
                || abs(chunk.getPosition().y() - cz) > radius + 1
            ) {
                toDelete.add(chunk)
            }
        }
        toDelete.forEach { obj -> obj.delete() }
        this.chunks.removeAll(toDelete)

        val chunks: MutableList<Chunk> = ArrayList<Chunk>()

        for (x in cx - radius..<cx + radius) {
            for (z in cz - radius..<cz + radius) {
                if (!hasChunk(x, z)) chunks.add(Chunk(this, x, z))
            }
        }

        chunks.sortBy { c -> c.getPosition().distanceSquared(cx, cz).toInt() }

        this.chunks.addAll(chunks)

        val futures = chunks.map { chunk ->
            CompletableFuture
                .runAsync(Runnable { chunk.generate(this.generator) }, this.executorService)
                .exceptionally(Function { f: Throwable ->
                    System.err.println(
                        "Failed to generate chunk " +
                                chunk.getPosition().x() + ", " +
                                chunk.getPosition().y() + " "
                    )
                    f.printStackTrace()
                    null
                })
        }

        return CompletableFuture.allOf(*futures.toTypedArray()).thenRun(
            Runnable {
                chunks.flatMap { c ->
                    listOf(
                        c as IChunk,
                        getChunk(c.getPosition().x(), c.getPosition().y() + 1),
                        getChunk(c.getPosition().x(), c.getPosition().y() - 1),
                        getChunk(c.getPosition().x() + 1, c.getPosition().y()),
                        getChunk(c.getPosition().x() - 1, c.getPosition().y())
                    )
                }.distinct().forEach { obj -> obj.updateMesh() }
            })
    }

    companion object {
        fun worldToChunkCoordinate(w: Int): Int {
            return if (w >= 0) w / ChunkConstants.SIZE else (w + 1) / ChunkConstants.SIZE - 1
        }
    }
}
